pkg/webhook/mutation.go

```diff
@@ -0,0 +1,184 @@
+
+	AddToManagerFuncs = append(AddToManagerFuncs, AddMutatingWebhook)
+
+	if err := apis.AddToScheme(runtimeScheme); err != nil {
+		log.Error(err, "unable to add to scheme")
+		panic(err)
+	}
+}
+
+// +kubebuilder:webhook:verbs=create,path=/v1/mutate,mutating=true,failurePolicy=ignore,groups=*,resources=*,versions=*,name=mutation.gatekeeper.sh
+// +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;update
+
+// AddMutatingWebhook registers the mutating webhook server with the manager
+func AddMutatingWebhook(mgr manager.Manager, client *opa.Client, processExcluder *process.Excluder) error {
+	if !*MutationEnabled {
+		return nil
+	}
+	reporter, err := newStatsReporter()
+	if err != nil {
+		return err
+	}
+	eventBroadcaster := record.NewBroadcaster()
+	kubeClient := kubernetes.NewForConfigOrDie(mgr.GetConfig())
+
+	eventBroadcaster.StartRecordingToSink(&clientcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(
+		scheme.Scheme,
+		corev1.EventSource{Component: "gatekeeper-mutation-webhook"})
+
+	wh := &admission.Webhook{
+		Handler: &mutationHandler{
+			webhookHandler: webhookHandler{
+				client:          mgr.GetClient(),
+				reader:          mgr.GetAPIReader(),
+				reporter:        reporter,
+				processExcluder: processExcluder,
+				eventRecorder:   recorder,
+				gkNamespace:     util.GetNamespace(),
+			},
+		},
+	}
+
+	// TODO(https://github.com/open-policy-agent/gatekeeper/issues/661): remove log injection if the race condition in the cited bug is eliminated.
+	// Otherwise we risk having unstable logger names for the webhook.
+	if err := wh.InjectLogger(log); err != nil {
+		return err
+	}
+	mgr.GetWebhookServer().Register("/v1/mutate", wh)
+
+	return nil
+}
+
+var _ admission.Handler = &mutationHandler{}
+
+type mutationHandler struct {
+	webhookHandler
+}
+
+// Handle the validation request
+func (h *mutationHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := log.WithValues("hookType", "mutation")
+	var timeStart = time.Now()
+
+	if isGkServiceAccount(req.AdmissionRequest.UserInfo) {
+		return admission.ValidationResponse(true, "Gatekeeper does not self-manage")
+	}
+
+	if req.AdmissionRequest.Operation != admissionv1beta1.Create &&
+		req.AdmissionRequest.Operation != admissionv1beta1.Update {
+		return admission.ValidationResponse(true, "Mutating only on create")
+	}
+
+	if h.isGatekeeperResource(ctx, req) {
+		return admission.ValidationResponse(true, "Not mutating gatekeeper resources")
+	}
+
+	requestResponse := mutationUnknownResponse
+	defer func() {
+		if h.reporter != nil {
+			if err := h.reporter.ReportMutationRequest(
+				requestResponse, time.Since(timeStart)); err != nil {
+				log.Error(err, "failed to report request")
+			}
+		}
+	}()
+
+	// namespace is excluded from webhook using config
+	if h.skipExcludedNamespace(req.AdmissionRequest.Namespace) {
```

Comment:
We may want to have a separate exclusion process for mutation webhooks, though that's certainly worth a debate

Reply:
Let me open a new PR with separate processes for validating/mutating webhooks, so we can have the discussion there.
This would also need a change in the config, to allow for separate configuration.